drivers/net/ethernet/intel/idpf/idpf_eth_idc.c

```diff
@@ -0,0 +1,453 @@
+		new_eth_dev = adapter->adevs[i];
+		if (!new_eth_dev)
+			goto alloc_exit;
+
+		new_eth_dev->adev.id = ida_alloc(&idpf_eth_idc_ida, GFP_KERNEL);
+		new_eth_dev->eth_info.eth_shared = &adapter->eth_shared;
+		new_eth_dev->eth_info.idx = i;
+
+		/* Set vport type */
+		new_eth_dev->eth_info.vport_type = vport_type;
+
+		if (vport_type == IDPF_DEFAULT_VPORT)
+			queue_type = IDPF_QUEUE_DEFAULT_VPORT;
+		else
+			queue_type = IDPF_QUEUE_DYNAMIC_VPORT;
+
+		dev_info = &new_eth_dev->eth_info;
+		err = idpf_alloc_max_qs(adapter, &dev_info->caps.q_info,
+					queue_type);
+		if (err) {
+			kfree(adapter->adevs[i]);
+			adapter->adevs[i] = NULL;
+			new_eth_dev = NULL;
+		}
+
+		return new_eth_dev;
+	}
+
+alloc_exit:
+	return NULL;
+}
+
+/**
+ * idpf_eth_idc_device_init - Initialize ethernet IDC & device(s)
+ * @adapter: Idpf private structure
+ */
+void idpf_eth_idc_device_init(struct idpf_adapter *adapter)
+{
+	struct device *dev = &adapter->eth_shared.pdev->dev;
+	struct idpf_eth_idc_auxiliary_driver *eth_idc_drv;
+	struct idpf_eth_idc_auxiliary_dev *eth_dev;
+	int err;
+	u16 i;
+
+	if (adapter->adevs) {
+		/* This is reset case so simply invoke add explicitly.
+		 * Direct eth device re-add or probe
+		 */
+		eth_idc_drv = idpf_eth_idc_get_driver();
+		for (i = 0; i < adapter->default_vports; ++i) {
+			eth_dev = adapter->adevs[i];
+			if (!eth_dev)
+				continue;
+
+			eth_idc_drv->adrv.probe(&eth_dev->adev, NULL);
+		}
+		idpf_eth_idc_dispatch_event(adapter,
+					    IDPF_ETH_IDC_EVENT_ALL_VPORTS,
+					    IDPF_ETH_IDC_EVENT_PROBE_POST_INIT,
+					    NULL);
+		return;
+	}
+
+	/* Following is the case of initial probe */
+	adapter->adevs =
+		kzalloc((adapter->max_vports *
+			 sizeof(struct idpf_eth_idc_auxiliary_dev *)),
+			GFP_KERNEL);
+	if (!adapter->adevs)
+		return;
+
+	for (i = 0; i < adapter->default_vports; ++i) {
+		eth_dev = idpf_eth_idc_device_alloc(adapter,
+						    IDPF_DEFAULT_VPORT);
+		if (!eth_dev)
+			return;
+		eth_dev->adev.name = "eth";
+		eth_dev->adev.dev.release = idpf_eth_idc_device_free;
+		eth_dev->adev.dev.parent =
+			idpf_eth_shared_to_dev(&adapter->eth_shared);
+
+		/* Initialize ethernet parameter(s) */
+		idpf_eth_idc_init_device_params(adapter, eth_dev);
+
+		err = auxiliary_device_init(&eth_dev->adev);
+		if (err) {
+			dev_err(dev, "Auxiliary dev ID 0x%x init failed 0x%x\n",
+				i, err);
+		} else {
+			err = auxiliary_device_add(&eth_dev->adev);
+			if (err)
+				continue;
+		}
+	}
+	idpf_eth_idc_dispatch_event(adapter,
+				    IDPF_ETH_IDC_EVENT_ALL_VPORTS,
+				    IDPF_ETH_IDC_EVENT_PROBE_POST_INIT,
+				    NULL);
+}
+
+/**
+ * idpf_eth_idc_device_deinit - Deinitialize ethernet IDC
+ * @adapter: Idpf private structure
+ */
+void idpf_eth_idc_device_deinit(struct idpf_adapter *adapter)
+{
+	u16 i;
+
+	for (i = 0; i < adapter->default_vports; ++i) {
+		auxiliary_device_delete(&adapter->adevs[i]->adev);
+		auxiliary_device_uninit(&adapter->adevs[i]->adev);
+	}
+
+	kfree(adapter->adevs);
```

Comment:
There is one race condition here. Suppose all the 'release' functions have not finished executing and we free the adevs here, that would be a problem. Please check how OOT is handling it. 
When I say 'release' function, I assume you are using a aux release function similar to OOT 
aux_dev->adev.dev.release = idpf_aux_device_free;